src/python/torchdistx/slow_momentum/slow_momentum_optimizer.py

```diff
@@ -0,0 +1,212 @@
+        slowmo_freq: Specifies how often (number of iterations) slow momentum
+            is to be performed (default: 48)
+        slowmo_factor: This specifies the value of slowmo momentum
+            to be used (default: 0.5)
+        slowmo_lr: This specifies the value of slowmo learning rate
+            to be used (default: 1.0)
+
+    Example::
+
+        >>>  import torch
+        >>>  import torch.distributed as dist
+        >>>  from torch.distributed.fsdp import(
+        >>>    FullyShardedDataParallel as FSDP
+        >>>  )
+        >>>  from torchdistx.slow_momentum import(
+        >>>     slow_momentum_comm,
+        >>>     slow_momentum_optimizer
+        >>>  )
+        >>>
+        >>>  net = torch.nn.Linear(4,10)
+        >>>  fsdp_net = FSDP(net)
+        >>>
+        >>>  # This implementation communicate gradients between
+        >>>  # workers of the same node,
+        >>>  # before averaging of model's parameters between nodes.
+        >>>  # The following creates intra-node subgroups
+        >>>  # and SlowMoState will take care of storing all required
+        >>>  # parameters for intra-node communication,
+        >>>  # i.e. pre- and post-division factors, and subgroups.
+        >>>  # To disable any communication between workers,
+        >>>  # set `grad_sync` to `False`
+        >>>  cur_subgroup, _ = dist.new_subgroups()
+        >>>  slowMoState = slow_momentum_comm.SlowMoState(
+        >>>     cur_subgroup,
+        >>>     grad_sync=True
+        >>>  )
+        >>>
+        >>>  # Register SlowMo hook, which only communicates gradients
+        >>>  # in a intra-node fashion.
+        >>>  fsdp_net.register_comm_hook(
+        >>>     slowMoState,
+        >>>     slow_momentum_comm.slowmo_hook
+        >>>  )
+        >>>
+        >>>  base_optimizer = torch.optim.SGD(
+        >>>     fsdp_net_slowmo.parameters(),
+        >>>     lr=1e-2
+        >>>  )
+        >>>  # Create a SlowMo optimizer that wraps a local optimizer.
+        >>>  slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+        >>>     base_optim=base_optimizer,
+        >>>     slowmo_freq=6,
+        >>>     slowmo_factor=0.5,
+        >>>     slowmo_lr=0.1
+        >>>  )
+        >>>
+        >>>  # SlowMo runs intra-node gradient averaging at every step,
+        >>>  # every 6th step it will run model averaging and
+        >>>  # a slow momentum update.
+        >>>  for step in range(0, 200):
+        >>>     slowmo_optim.zero_grad()
+        >>>     loss = loss_fn(output, labels)
+        >>>     loss.backward()
+        >>>     slowmo_optim.step()
+    """
+
+    def __init__(
+        self,
+        base_optim: torch.optim.Optimizer,
+        slowmo_freq: int = 48,
+        slowmo_factor: float = 0.5,
+        slowmo_lr: float = 1.0,
+    ):
+        if base_optim is None:
+            raise ValueError("Base optimizer is a required parameter.")
```

Comment:
I agree with 
> At some point, we should not gracefully error if the user blatantly fails to follow the API.

I think, I will keep it here this time, since if base_optim is `None`, next checks will fail and it may be not straightforward to a user, why (not everyone, of course).